### ISSUE TITLE
Cannot start local dev galaxy: recursive_guard

### ISSUE THREAD

**Issue by hexylena** [galaxyproject/galaxy#18849]
I'm on 24.0.2 with minimal changes

```
$ git status | head
HEAD detached at v24.0.2
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   config/plugins/webhooks/gtn/script.js
        modified:   config/plugins/webhooks/gtn/styles.css

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        0001-can-it-be-that-easy.patch
```

I have deleted my .venv folder and allowed that to be recreated, but every start results in the same traceback:

```
$ GALAXY_CONFIG_FILE=config/galaxy.yml sh run.sh                                                                     
Activating virtualenv at .venv                                                                                                                                              
Requirement already satisfied: pip>=20.3 in ./.venv/lib64/python3.12/site-packages (23.2.1)                                                                                 
Requirement already satisfied: wheel in ./.venv/lib64/python3.12/site-packages (0.44.0)                                                                                                                                                                                                                                                                 
[notice] A new release of pip is available: 23.2.1 -> 24.2                                                                                                                  
[notice] To update, run: pip install --upgrade pip        
...
[notice] A new release of pip is available: 23.2.1 -> 24.2                                                                                                                  
[notice] To update, run: pip install --upgrade pip                  
The Galaxy client build is up to date and will not be rebuilt at this time.                                                                                                 
Activating virtualenv at .venv                                                        
Traceback (most recent call last):                                                                                                                                          
  File "/home/user/arbeit/galaxy/galaxy/.venv/bin/galaxyctl", line 8, in <module>                                                                                           
    sys.exit(galaxyctl())                                                                                                                                                   
             ^^^^^^^^^^^                                                              
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/click/core.py", line 1157, in __call__                        
    return self.main(*args, **kwargs)                                                 
           ^^^^^^^^^^^^^^^^^^^^^^^^^^                                                 
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/click/core.py", line 1078, in main                                                             
    rv = self.invoke(ctx)                                                             
         ^^^^^^^^^^^^^^^^                                                             
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/click/core.py", line 1682, in invoke  
    cmd_name, cmd, args = self.resolve_command(ctx, args)                                                                                                                   
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                   
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/click/core.py", line 1729, in resolve_command
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                  11:35:01 [38/545]
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/click/core.py", line 1729, in resolve_command         
    cmd = self.get_command(ctx, cmd_name)                                                                                                                                   
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                   
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/gravity/cli.py", line 59, in get_command                                
    return name_to_command(name)                                                                                                                                            
           ^^^^^^^^^^^^^^^^^^^^^                                                      
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/gravity/cli.py", line 45, in name_to_command                                                   
    mod = __import__(mod_name, None, None, ["cli"])                                   
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                         
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/gravity/commands/cmd_update.py", line 3, in <module>                                           
    from gravity import process_manager                                                                                                                                     
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/gravity/process_manager/__init__.py", line 14, in <module>                                     
    from gravity.config_manager import ConfigManager                                                                                                                        
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/gravity/config_manager.py", line 18, in <module>                                               
    from gravity.state import ConfigFile, service_for_service_type                                                                                                          
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/gravity/state.py", line 42, in <module>                                                        
    class ConfigFile(BaseModel):                                                                                                                                            
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/pydantic/v1/main.py", line 286, in __new__                                                     
    cls.__try_update_forward_refs__()                                                                                                                                       
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/pydantic/v1/main.py", line 808, in __try_update_forward_refs__                                 
    update_model_forward_refs(cls, cls.__fields__.values(), cls.__config__.json_encoders, localns, (NameError,))                                                            
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/pydantic/v1/typing.py", line 554, in update_model_forward_refs                                 
    update_field_forward_refs(f, globalns=globalns, localns=localns)
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/pydantic/v1/typing.py", line 520, in update_field_forward_refs                                 
    field.type_ = evaluate_forwardref(field.type_, globalns, localns or None)         
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                               
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/pydantic/v1/typing.py", line 66, in evaluate_forwardref                                        
    return cast(Any, type_)._evaluate(globalns, localns, set())                                                                                                             
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                       
TypeError: ForwardRef._evaluate() missing 1 required keyword-only argument: 'recursive_guard'                                              
Executing: galaxy                                                                     
Traceback (most recent call last):                                                    
  File "/home/user/arbeit/galaxy/galaxy/.venv/bin/galaxy", line 8, in <module>                                                                                              
    sys.exit(galaxy())                                                                
             ^^^^^^^^                                                                 
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)                                                                                                                                       
           ^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                       
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/click/core.py", line 1078, in main           
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/click/core.py", line 1434, in invoke                                    
    return ctx.invoke(self.callback, **ctx.params)                                                                                                                          
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                    
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/click/core.py", line 783, in invoke                                                            
    return __callback(*args, **kwargs)                                                
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                      
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/click/decorators.py", line 33, in new_func                                                     
    return f(get_current_context(), *args, **kwargs)                                                                                                                        
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                        
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/gravity/cli.py", line 77, in galaxy                                                            
    mod = __import__("gravity.commands.cmd_start", None, None, ["cli"])                                                                                                     
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                     
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/gravity/commands/cmd_start.py", line 4, in <module>                                            
    from gravity import process_manager                                                                                                                                     
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/gravity/process_manager/__init__.py", line 14, in <module>                                     
    from gravity.config_manager import ConfigManager                                                                                                                        
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/gravity/config_manager.py", line 18, in <module>                                               
    from gravity.state import ConfigFile, service_for_service_type                                                                                                          
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/gravity/state.py", line 42, in <module>                                                        
    class ConfigFile(BaseModel):                                                      
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/pydantic/v1/main.py", line 286, in __new__                                                     
    cls.__try_update_forward_refs__()                                                 
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/pydantic/v1/main.py", line 808, in __try_update_forward_refs__                                 
    update_model_forward_refs(cls, cls.__fields__.values(), cls.__config__.json_encoders, localns, (NameError,))                                                            
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/pydantic/v1/typing.py", line 554, in update_model_forward_refs                                 
    update_field_forward_refs(f, globalns=globalns, localns=localns)                  
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/pydantic/v1/typing.py", line 520, in update_field_forward_refs
    field.type_ = evaluate_forwardref(field.type_, globalns, localns or None)         
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^         
  File "/home/user/arbeit/galaxy/galaxy/.venv/lib64/python3.12/site-packages/pydantic/v1/typing.py", line 66, in evaluate_forwardref                                        
    return cast(Any, type_)._evaluate(globalns, localns, set())                       
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                       
TypeError: ForwardRef._evaluate() missing 1 required keyword-only argument: 'recursive_guard'                      
GALAXY_CONFIG_FILE=config/galaxy.yml sh run.sh  22.43s user 1.00s system 97% cpu 23.914 total                                                                               
```

**Comment by bgruening** [comment]
Your Python version is too recent. You need 3.11 afaik for the time being. I have seen this on CI as well.

**Comment by nsoranzo** [comment]
Indeed, see e.g.: https://github.com/galaxyproject/galaxy/actions/runs/10794294798/job/29939507503
Galaxy 24.1 should work fine though on Python 3.12, at least it does on CI, let us know if it doesn't for you.

**Comment by hexylena** [comment]
Urgh, not something I have a lot of control over. 

It would be very helpful if it could exit with a more useful error code. We know the permitted versions of python, couldn't such a check be placed very, very early in startup? Or run.sh?

**Comment by nsoranzo** [comment]
@hexylena The problem is that this broke with the release of Python 3.12.4 (see https://github.com/pydantic/pydantic/issues/9637 ), which was released *after* Galaxy 24.0; previous versions of Python 3.12 worked fine.
The issue was fixed in pydantic 2.7.4 (see https://github.com/pydantic/pydantic/blob/main/HISTORY.md#v274-2024-06-12 ) but this is way too new to be used on Galaxy 24.0 (which is using the pydantic 2.5.x series).

The best workaround I can think of is to open a PR against release_24.0 to:
- amend `scripts/check_python.py` to bail out if Python >3.12.3
- modify the failing GitHub workflows to use 3.12.3 as `python-version` instead of 3.12

Both of these changes would need to be reverted when merging forward.
Not sure if this is worth the effort given that 24.1 is our latest release and doesn't have the issue.
